src/Confluent.Kafka.Avro/GenericDeserializerImpl.cs

```diff
@@ -58,31 +56,23 @@ public GenericRecord Deserialize(string topic, byte[] array)
                 }
                 var writerId = IPAddress.NetworkToHostOrder(reader.ReadInt32());
 
-                DatumReader<GenericRecord> datumReader = null;
-
-                lock (datumReaderLock)
+                datumReaderBySchemaId.TryGetValue(writerId, out DatumReader<GenericRecord> datumReader);
```

Comment:
There is a race condition between this TryGetValue and the insert below.

Reply:
I think it's benign. We don't care if we have make more than one datumReader - any of them will be valid - as long as we definitely have a valid instance to do the read (which we do). Ideally we wouldn't ever create more than one datumReader for a given writerId (hence what I was trying to do with the longer lasting lock), but it doesn't matter if we do (except possibly from a perf perspective), and as you point out, concurrent consuming won't be very common. Another thing that can happen (will be very uncommon) is the cache is cleared that includes a datumReader that's going to be used by another thread, but again that doesn't matter - the only thing that really matters is we have a valid datumReader instance when we go to use it.

Reply:
If this is indeed intentional make a very clear comment about it, because this looks like a mistake/bug.

Reply:
ok, will add a note.